packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -332,6 +332,74 @@ abstract class ImageProvider<T> {
     return ImageStream();
   }
 
+  /// Checks if this [ImageProvider] has a matching key in the
+  /// [PaintingBinding.imageCache]
+  ///
+  /// The `cache` and `configuration` parameters must not be null. If the
+  /// `handleError` parameter is null, errors will be reported to
+  /// [FlutterError.onError], and the method will return null.
+  ///
+  /// A completed return value of null indicates that an error has occurred.
+  Future<ImageCacheLocation> findCacheLocation({
+    @required ImageConfiguration configuration,
+    ImageErrorListener handleError,
+  }) {
+    assert(configuration != null);
+    final Completer<ImageCacheLocation> completer = Completer<ImageCacheLocation>();
+    bool didError = false;
+    final ImageErrorListener errorHandler = (dynamic exception, StackTrace stack) {
+      if (didError) {
+        return;
+      }
+      didError = true;
+      if (handleError != null) {
+        handleError(exception, stack);
+      } else {
+        FlutterError.onError(FlutterErrorDetails(
+          context: ErrorDescription('while checking the cache location of an image'),
+          informationCollector: () sync* {
+            yield DiagnosticsProperty<ImageProvider>('Image provider', this);
+            yield DiagnosticsProperty<ImageConfiguration>('Image configuration', configuration);
+          },
+          exception: exception,
+          stack: stack,
+        ));
+        completer.complete(null);
+      }
+    };
+
+
+    // If an error is added to a synchronous completer before a listener has been
+    // added, it can throw an error both into the zone and up the stack. Thus, it
+    // looks like the error has been caught, but it is in fact also bubbling to the
+    // zone. Since we cannot prevent all usage of Completer.sync here, or rather
+    // that changing them would be too breaking, we instead hook into the same
+    // zone mechanism to intercept the uncaught error and deliver it to the
+    // supplied error handler. Note that these errors may be duplicated,
+    // hence the need for the `didError` flag.
+    final Zone dangerZone = Zone.current.fork(
+      specification: ZoneSpecification(
+        handleUncaughtError: (Zone zone, ZoneDelegate delegate, Zone parent, Object error, StackTrace stackTrace) {
```

Comment:
Drive-by comment feel free to ignore especially since this is an pre-existing issue with the ImageProvider (which I wanted to file a bug for the longest time). It is not safe for flutter to use Dart zones with error handlers.
Dart's default Future implementations have a special case that errors cannot leave a zone with an error handler and the future will just never fulfill meaning that some continuations might never be executed. It will appear to work while you don't have errors, once you have an error you can kill the users control flow in a hard to understand and impossible to handle manner.
(Dart folks are aware of the issue but changing/fixing the behavior is afaik not currently on the roadmap, meanwhile we're kinda a stuck with use error zones in your main() function and nowhere else policy)

Reply:
/cc @jonahwilliams

Reply:
Jonah knows more about this - I'm copying the implementation here from the other current implementation.  If there's a bug in that we should file it and fix it separately from here.